pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,951 @@
+		internalLocal := svc.Spec.InternalTrafficPolicy != nil && *svc.Spec.InternalTrafficPolicy == corev1.ServiceInternalTrafficPolicyLocal
+		externalLocal := svc.Spec.ExternalTrafficPolicy == corev1.ServiceExternalTrafficPolicyLocal
+		var hasLocalEndpoints bool
+		if internalLocal || externalLocal {
+			hasLocalEndpoints = c.hasLocalEndpoints(svc)
+		}
+		if ipTypes.Has(v1alpha1.ServiceIPTypeClusterIP) {
+			if internalLocal && hasLocalEndpoints || !internalLocal {
+				for _, clusterIP := range svc.Spec.ClusterIPs {
+					serviceIPs = append(serviceIPs, clusterIP)
+				}
+			}
+		}
+		if ipTypes.Has(v1alpha1.ServiceIPTypeExternalIP) {
+			if externalLocal && hasLocalEndpoints || !externalLocal {
+				for _, externalIP := range svc.Spec.ExternalIPs {
+					serviceIPs = append(serviceIPs, externalIP)
+				}
+			}
+		}
+		if ipTypes.Has(v1alpha1.ServiceIPTypeLoadBalancerIP) && svc.Spec.Type == corev1.ServiceTypeLoadBalancer {
+			if externalLocal && hasLocalEndpoints || !externalLocal {
+				serviceIPs = append(serviceIPs, getIngressIPs(svc)...)
+			}
+		}
+	}
+
+	for _, ip := range serviceIPs {
+		if c.enabledIPv4 && utilnet.IsIPv4String(ip) {
+			allRoutes.Insert(bgp.Route{Prefix: ip + ipv4Suffix})
+		}
+		if c.enabledIPv6 && utilnet.IsIPv6String(ip) {
+			allRoutes.Insert(bgp.Route{Prefix: ip + ipv6Suffix})
+		}
+	}
+}
+
+func (c *Controller) addEgressRoutes(allRoutes sets.Set[bgp.Route]) {
+	egresses, _ := c.egressLister.List(labels.Everything())
+	for _, eg := range egresses {
+		if eg.Status.EgressNode != c.nodeName {
+			continue
+		}
+		ip := eg.Status.EgressIP
+		if c.enabledIPv4 && utilnet.IsIPv4String(ip) {
+			allRoutes.Insert(bgp.Route{Prefix: ip + ipv4Suffix})
+		}
+		if c.enabledIPv6 && utilnet.IsIPv6String(ip) {
+			allRoutes.Insert(bgp.Route{Prefix: ip + ipv6Suffix})
+		}
+	}
+}
+
+func (c *Controller) addPodRoutes(allRoutes sets.Set[bgp.Route]) {
+	if c.enabledIPv4 {
+		allRoutes.Insert(bgp.Route{Prefix: c.podIPv4CIDR})
+	}
+	if c.enabledIPv6 {
+		allRoutes.Insert(bgp.Route{Prefix: c.podIPv6CIDR})
+	}
+}
+
+func (c *Controller) hasLocalEndpoints(svc *corev1.Service) bool {
+	labelSelector := labels.Set{discovery.LabelServiceName: svc.GetName()}.AsSelector()
+	items, _ := c.endpointSliceLister.EndpointSlices(svc.GetNamespace()).List(labelSelector)
+	for _, eps := range items {
+		for _, ep := range eps.Endpoints {
+			if ep.NodeName != nil && *ep.NodeName == c.nodeName {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (c *Controller) getPeerConfigs(peers []v1alpha1.BGPPeer) map[string]bgp.PeerConfig {
+	c.bgpPeerPasswordsMutex.RLock()
+	defer c.bgpPeerPasswordsMutex.RUnlock()
+
+	peerConfigs := make(map[string]bgp.PeerConfig)
+	for i := range peers {
+		if c.enabledIPv4 && utilnet.IsIPv4String(peers[i].Address) ||
+			c.enabledIPv6 && utilnet.IsIPv6String(peers[i].Address) {
+			peerKey := generateBGPPeerKey(peers[i].Address, peers[i].ASN)
+
+			var password string
+			if p, exists := c.bgpPeerPasswords[peerKey]; exists {
```

Comment:
I am a little confused about the Secret, who will create and update this Secret `antrea-bgp-passwords`?

Reply:
OK, so I suppose the creation of this Secret should be a part of the usage document? I didn't see you mention this in #6524

Reply:
Added corresponding paragraph in #6524.